File: java/src/main/java/io/github/emindeniz99/applepurchasereceiptverifier/receipt/InAppPurchase.java
package io.github.emindeniz99.applepurchasereceiptverifier.receipt;

import java.time.Instant;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import org.jspecify.annotations.Nullable;

/**
 * One in-app purchase from a legacy app receipt (attribute 17). Field set
 * per Apple's "Validating receipts on the device" attribute table;
 * {@code null} means the attribute was absent from the receipt.
 */
public final class InAppPurchase {

    private final @Nullable Long quantity;
    private final @Nullable String productId;
    private final @Nullable String transactionId;
    private final @Nullable String originalTransactionId;
    private final @Nullable Instant purchaseDate;
    private final @Nullable Instant originalPurchaseDate;
    private final @Nullable Instant expiresDate;
    private final @Nullable Instant cancellationDate;
    private final @Nullable Long webOrderLineItemId;
    private final @Nullable Long isTrialPeriod;
    private final @Nullable Long isInIntroOfferPeriod;
    private final Map<Integer, List<byte[]>> unknownAttributes;

    InAppPurchase(
            @Nullable Long quantity,
            @Nullable String productId,
            @Nullable String transactionId,
            @Nullable String originalTransactionId,
            @Nullable Instant purchaseDate,
            @Nullable Instant originalPurchaseDate,
            @Nullable Instant expiresDate,
            @Nullable Instant cancellationDate,
            @Nullable Long webOrderLineItemId,
            @Nullable Long isTrialPeriod,
            @Nullable Long isInIntroOfferPeriod,
            Map<Integer, List<byte[]>> unknownAttributes) {
        this.quantity = quantity;
        this.productId = productId;
        this.transactionId = transactionId;
        this.originalTransactionId = originalTransactionId;
        this.purchaseDate = purchaseDate;
        this.originalPurchaseDate = originalPurchaseDate;
        this.expiresDate = expiresDate;
        this.cancellationDate = cancellationDate;
        this.webOrderLineItemId = webOrderLineItemId;
        this.isTrialPeriod = isTrialPeriod;
        this.isInIntroOfferPeriod = isInIntroOfferPeriod;
        this.unknownAttributes = Collections.unmodifiableMap(unknownAttributes);
    }

    /** Number of units purchased (attribute 1701). */
    public @Nullable Long quantity() {
        return quantity;
    }

    public @Nullable String productId() {
        return productId;
    }

    public @Nullable String transactionId() {
        return transactionId;
    }

    public @Nullable String originalTransactionId() {
        return originalTransactionId;
    }

    /** Purchase date (attribute 1704), an {@link Instant}, not epoch milliseconds. */
    public @Nullable Instant purchaseDate() {
        return purchaseDate;
    }

    /** Purchase date of the original transaction (attribute 1706), an {@link Instant}. */
    public @Nullable Instant originalPurchaseDate() {
        return originalPurchaseDate;
    }

    /** Subscription expiration (attribute 1708), if this is a subscription. */
    public @Nullable Instant expiresDate() {
        return expiresDate;
    }

    /** Set when Apple customer support cancelled/refunded (attribute 1712). */
    public @Nullable Instant cancellationDate() {
        return cancellationDate;
    }

    public @Nullable Long webOrderLineItemId() {
        return webOrderLineItemId;
    }

    /**
     * Attribute 1713 — 1 while the purchase is inside a free trial, 0
     * otherwise. Carried as an integer like
     * {@link #isInIntroOfferPeriod()}, which Apple's verifyReceipt answer
     * renders as the string "true"/"false".
     */
    public @Nullable Long isTrialPeriod() {
        return isTrialPeriod;
    }

    public @Nullable Long isInIntroOfferPeriod() {
        return isInIntroOfferPeriod;
    }

    /**
     * Raw unmodeled attributes by type, so a field Apple adds later is not
     * lost.
     *
     * <p>A fresh copy each call, arrays included, for the same reason as
     * {@link AppReceipt#unknownAttributes()}: the shared {@code byte[]}
     * contents could otherwise be rewritten by one caller for every other.
     */
    public Map<Integer, List<byte[]>> unknownAttributes() {
        return AppReceipt.deepCopy(unknownAttributes);
    }
}

File: php/src/Receipt/InAppPurchase.php
<?php

declare(strict_types=1);

namespace EminDeniz99\ApplePurchaseReceiptVerifier\Receipt;

use DateTimeImmutable;

/**
 * One in-app purchase from a verified legacy app receipt (attribute 17).
 *
 * Dates are `DateTimeImmutable` in UTC — receipt attributes carry RFC 3339
 * text, so a native date type is the honest representation. (JWS payload
 * claims are the opposite case: Apple ships those as epoch-millisecond
 * integers and they stay integers — see {@see \EminDeniz99\ApplePurchaseReceiptVerifier\Jws\TransactionPayload}.)
 */
final class InAppPurchase
{
    /**
     * Properties whose value is raw bytes rather than text. PHP spells both
     * `string`, so this is how a generic consumer (a serializer, a debug
     * dumper, the conformance adapter) knows to hex-encode rather than print.
     */
    public const BINARY_PROPERTIES = ['unknownAttributes'];

    /**
     * @param int|null $isTrialPeriod attribute 1713 — 1 while the purchase is
     *        inside a free trial, 0 otherwise. Carried as an integer like
     *        `isInIntroOfferPeriod`, which Apple's verifyReceipt answer
     *        renders as the string "true"/"false". On none of Apple's pages;
     *        established by comparing a genuine production receipt with
     *        Apple's verifyReceipt answer (measured 2026-09-21).
     * @param array<int, list<string>> $unknownAttributes raw values of attribute
     *        types this library does not model, keyed by type — forward
     *        compatibility for fields Apple may add (PLAN.md D10). Verified,
     *        but undecoded.
     */
    public function __construct(
        public readonly ?int $quantity,
        public readonly ?string $productId,
        public readonly ?string $transactionId,
        public readonly ?string $originalTransactionId,
        public readonly ?DateTimeImmutable $purchaseDate,
        public readonly ?DateTimeImmutable $originalPurchaseDate,
        public readonly ?DateTimeImmutable $expiresDate,
        public readonly ?DateTimeImmutable $cancellationDate,
        public readonly ?int $webOrderLineItemId,
        public readonly ?int $isTrialPeriod,
        public readonly ?int $isInIntroOfferPeriod,
        public readonly array $unknownAttributes,
    ) {
    }
}
